quic/s2n-quic-transport/src/connection/api.rs

```diff
@@ -201,3 +213,130 @@ impl Connection {
         self.api.query_event_context_mut(query)
     }
 }
+
+/// An opaque token issued to each connection handle which allows the stream
+/// controller to track any pending open requests.
+///
+/// Each connection handle must have a unique instance of a token so their wakers are correctly
+/// tracked.
+#[derive(Debug)]
+pub struct OpenToken(Option<NonZeroU64>);
+
+impl OpenToken {
+    /// Creates a new open token
+    ///
+    /// This should be held on the connection handle and should be presented each time
+    /// `poll_open_stream` is called.
+    #[inline]
+    pub const fn new() -> Self {
+        Self(None)
+    }
+
+    /// Creates an open token counter, used for issuing tokens to callers
+    #[inline]
+    pub(crate) fn counter() -> Self {
+        Self(Some(unsafe {
+            // Safety: 1 is always non-zero
+            NonZeroU64::new_unchecked(1)
+        }))
+    }
+
+    /// Returns the next open token for a connection
+    #[inline]
+    pub(crate) fn next(&mut self) -> Self {
+        let v = self.0.expect("token counter should always be initialized");
+        let next = Self(Some(v));
+        self.0 = Some(unsafe {
+            // Safety: N + 1 is always non-zero
+            NonZeroU64::new_unchecked(v.get() + 1)
+        });
+        next
+    }
+
+    /// Returns the index that the caller's waker should be stored, given the expired token
+    #[inline]
+    pub(crate) fn index(&self, expired_token: &Self) -> Option<usize> {
+        let base = expired_token.0.map_or(0, |v| v.get());
+        // since we're dealing with NonZeroU64, add one to get the correct index
+        let base = base + 1;
+        let v = self.0?;
+        let v = v.get().checked_sub(base)?;
+        Some(v as usize)
```

Comment:
why not just use `NonZeroUsize` since you're casting to usize?

Reply:
the cast to `usize` is after subtracting the expired tokens so it will fit even on 32-bit platforms. But since tokens are monotonic and an application can theoretically open 2^62 streams, we need to be explicit about it holding 64 bits.